### ISSUE TITLE
compile_fx inplace modifly the input graph

### ISSUE THREAD

**Issue by youkaichao** [pytorch/pytorch#138980]
### 🐛 Describe the bug

I'm compiling a graph multiple times using inductor. I find it inplace modify the graph, and one of the graph's output changes from tensor to list of tensor.

example code:

```python
import torch
from torch import nn

@torch.library.custom_op("silly::attention", mutates_args=["out"])
def silly_attention(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, out: torch.Tensor) -> None:
    print("silly")
    out.copy_(q)
    out[0] += 1

@silly_attention.register_fake
def _(q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, out: torch.Tensor) -> None:
    return

class SillyModel(nn.Module):
    def __init__(self) -> None:
        super().__init__()

    def forward(
        self,
        x: torch.Tensor
    ) -> torch.Tensor:
        x = x + 1
        x = x + 2
        out = torch.empty_like(x)
        torch.ops.silly.attention(x, x, x, out)
        x = out
        x = x - 2
        x = x - 1
        out = torch.empty_like(x)
        torch.ops.silly.attention(x, x, x, out)
        x = out
        x = x + 1
        return x

model = SillyModel()

def mybackend(graph, example_inputs):
    graph.print_readable()

    subgraph_id = 0
    node_to_subgraph_id = {}
    attention_graphs = []
    for node in graph.graph.nodes:
        if node.op in ("output", "placeholder"):
            continue
        if node.op == 'call_function' and str(node.target) in ["silly.attention"]:
            subgraph_id += 1
            node_to_subgraph_id[node] = subgraph_id
            attention_graphs.append(subgraph_id)
            subgraph_id += 1
        else:
            node_to_subgraph_id[node] = subgraph_id

    # `keep_original_order` is important!
    # otherwise pytorch might reorder the nodes and
    # the semantics of the graph will change when we
    # have mutations in the graph
    split_gm = torch.fx.passes.split_module.split_module(
        graph,
        None,
        lambda node: node_to_subgraph_id[node],
        keep_original_order=True)

    split_gm.print_readable()

    for (name, module) in list(split_gm.named_modules()):
        if name == "":
            # stitching module
            continue
        if "." in name:
            # recursive child module
            continue
        graph_id = int(name.replace("submod_", ""))
        if graph_id not in attention_graphs:
            # cannot setattr to a module, so we need to set it to the dict
            split_gm.__dict__[name] = piecewise_backend(
                module)

    print("first compilation")
    output = split_gm(*example_inputs)
    print(output.__class__)

    return split_gm.forward

def piecewise_backend(model):

    compiled_once = True
    compiled_twice = True
    compiled_graph_1 = None
    compiled_graph_2 = None
    def f(*args):
        model.print_readable()
        nonlocal compiled_once, compiled_twice, compiled_graph_1, compiled_graph_2
        if compiled_once:
            compiled_once = False
            from torch._inductor.compile_fx import compile_fx
            compiled_graph = compile_fx(model, args)
            return compiled_graph(*args)
        if compiled_twice:
            compiled_once = False
            from torch._inductor.compile_fx import compile_fx
            compiled_graph_1 = compile_fx(model, args)
        return compiled_graph_1(*args)
    return f

model = torch.compile(model, backend=mybackend)

input_buffer = torch.randn(100)
output = model(input_buffer)
print(output.__class__)

output = model(input_buffer)
print(output.__class__)
```

Some key output logging:

this is the last piece of graph:

```python
    class submod_4(torch.nn.Module):
        def forward(self, out_1: "f32[100]"):
             # File: /data/youkaichao/tmp/vllm/test_split.py:32 in forward, code: x = x + 1
            add: "f32[100]" = out_1 + 1;  out_1 = None
            return add
```

after calling `compile_fx`, it becomes:

```python
class GraphModule(torch.nn.Module):
    def forward(self, out_1: "f32[100]"):
         # File: /data/youkaichao/tmp/vllm/test_split.py:32 in forward, code: x = x + 1
        add: "f32[100]" = out_1 + 1;  out_1 = None
        return [add]
```

is it a desired behavior?

if I want to avoid it, how to clone a graph, without cloning the module and parameters, if I want to send a copy to `compile_fx`?

### Versions

pytorch 2.4 , 2.5, and 2.6.0.dev20241022+cu124

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @aakhundov

**Comment by youkaichao** [comment]
cc @Chillee

**Comment by youkaichao** [comment]
It seems adding `copy.deepcopy` solves the problem:

`compile_fx(copy.deepcopy(model), args)`

**Comment by desertfire** [comment]
Not sure what you exactly want for your use case, but I am closing this issue since deepcopy solves your problem. If you need more help, feel free to reopen.

**Comment by youkaichao** [comment]
I assume a compiler should not change its input, but inductor does inplace modification of its input. I think this still worth a fix. @desertfire

**Comment by youkaichao** [comment]
it becomes confusing for the following usage:

```python
out1 = compile_fx(graph)
out2 = compile_fx(graph)
# out1 is different from out2
```

**Comment by masnesral** [comment]
cc @desertfire to have another look per the last comment

**Comment by desertfire** [comment]
`compile_fx` is an internal API. It has certain expectation on the output format, and change it if it doesn't conform. Is there a reason you want to call that API directly?

**Comment by youkaichao** [comment]
> Is there a reason you want to call that API directly

yes, we want to call the inductor compiler directly, maybe many times, for one graph captured by dynamo.

**Comment by zou3519** [comment]
I think this is a wontfix, unless the compiler modularization effort wants to do something about it. @avikchaudhuri @ezyang.

In vLLM we are now using torch._inductor.standalone_compile, which does not in-place mutate the graph, so it avoids the deepcopy.

**Comment by karthickai** [comment]
**🤖 PT2 Repro Agent**

I tried to reproduce this issue on PyTorch nightly (`2.13.0.dev20260518+cu130`) on NVIDIA H100 and confirmed it reproduces (3/3 runs failed).

All 3 runs exit 1. Before the call under test, the input GraphModule's output-node args[0] is `add` (a torch.fx.Node); after the call it is `None` (NoneType). This confirms the input graph is mutated in place. The issue reports the mutation as `return tensor` becoming `return [tensor]`; on this nightly the post-call form is `(None,)` instead, but the same underlying in-place mutation of the input graph's output node is observed.

The compiler bisector narrowed the bug to the **inductor** backend; no single subsystem fully accounts for it (likely a multi-source issue accumulating across passes).

<details>
<summary>Error output</summary>

```
Before compile_fx: output args = (add,)
  type of args[0] = Node
After compile_fx: output args = (None,)
  type of args[0] = NoneType
Result: BUG REPRODUCED

ERROR conda.cli.main_run:execute(49): `conda run python repro.py` failed. (See above for error)
```
</details>

<details>
<summary>Repro script I used</summary>

```python
import os
import torch


def test_fn() -> bool:
    """Return True if NO bug (compile_fx does not mutate input graph).
    False if bug present (compile_fx mutates the output node of input graph)."""
    torch._dynamo.reset()
    backend = os.environ.get("TORCH_COMPILE_BACKEND", "inductor")
    device = "cuda" if torch.cuda.is_available() else "cpu"

    class M(torch.nn.Module):
        def forward(self, x):
            return x + 1

    m = M()

    if backend != "inductor":
        x = torch.randn(100, device=device)
        eager_out = m(x)
        compiled_out = torch.compile(m, backend=backend)(x)
        return torch.allclose(eager_out, compiled_out)

    traced = torch.fx.symbolic_trace(m)

    output_node = None
    for node in traced.graph.nodes:
        if node.op == "output":
            output_node = node
            break

    args_before = output_node.args
    output_was_single_tensor = isinstance(args_before[0], torch.fx.Node)

    from torch._inductor.compile_fx import compile_fx
    try:
        _ = compile_fx(traced, (torch.randn(100, device=device),))
    except Exception:
        pass

    args_after = output_node.args
    output_is_single_tensor = isinstance(args_after[0], torch.fx.Node)

    if output_was_single_tensor and not output_is_single_tensor:
        return False
    if args_before is not args_after:
        return False
    return True


if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"
    class _M(torch.nn.Module):
        def forward(self, x):
            return x + 1

    _traced = torch.fx.symbolic_trace(_M())
    _output_node = next(n for n in _traced.graph.nodes if n.op == "output")
    print(f"Before compile_fx: output args = {_output_node.args}")
    print(f"  type of args[0] = {type(_output_node.args[0]).__name__}")

    from torch._inductor.compile_fx import compile_fx as _compile_fx
    try:
        _ = _compile_fx(_traced, (torch.randn(100, device=device),))
    except Exception as e:
        print(f"compile_fx raised: {type(e).__name__}: {e}")

    print(f"After compile_fx: output args = {_output_node.args}")
    print(f"  type of args[0] = {type(_output_node.args[0]).__name__}")

    try:
        ok = test_fn()
    except Exception:
        import traceback
        traceback.print_exc()
        os._exit(2)
    print(f"Result: {'NO BUG' if ok else 'BUG REPRODUCED'}")
    os._exit(0 if ok else 1)

```
</details>

<details>
<summary>My environment</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260518+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.5.0 20240719 (Red Hat 11.5.0-14)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.13.12 | packaged by Anaconda, Inc. | (main, Feb 24 2026, 16:13:31) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.16.1-0_fbk2_hardened_0_gf40efc324cc8-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100
Nvidia driver version: 580.126.09
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  46
On-line CPU(s) list:                     0-45
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 9654 96-Core Processor
CPU family:                              25
Model:                                   17
Thread(s) per core:                      1
Core(s) per socket:                      46
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                4792.79
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm rep_good nopl cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp ibrs_enhanced vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr wbnoinvd arat npt lbrv nrip_save tsc_scale vmcb_clean flushbyasid pausefilter pfthreshold v_vmsave_vmload vgif vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid fsrm flush_l1d arch_capabilities
Virtualization:                          AMD-V
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               2.9 MiB (46 instances)
L1i cache:                               2.9 MiB (46 instances)
L2 cache:                                23 MiB (46 instances)
L3 cache:                                736 MiB (46 instances)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-45
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Vulnerable
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Vulnerable
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] onnx==1.21.0
[pip3] onnx-ir==0.2.1
[pip3] onnxscript==0.7.0
[pip3] torch==2.13.0.dev20260517+cu130
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0.dev20260517+cu130
[pip3] torchdata==0.11.0
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.9.0
[pip3] torchrec==1.7.0.dev20260413+cu130
[pip3] torchtune==0.6.1
[pip3] torchvision==0.28.0.dev20260518+cu130
[pip3] triton==3.7.0+git88b227e2
[conda] Could not collect
```
</details>

---
If the repro agent produced an incorrect result (wrong classification, bad repro script, missed the actual bug, etc.), add the label `repro:agent:h100:incorrect` to the issue.
